File: LeetCode/Median of Two Sorted Arrays/Median of Two Sorted Arrays.java
// Question : https://leetcode.com/problems/median-of-two-sorted-arrays/

// Example :
/**
 * Example 1:
 * Input: nums1 = [1,3], nums2 = [2]
 * Output: 2.00000
 * Explanation: merged array = [1,2,3] and median is 2.
 * 
 * Example 2:
 * Input: nums1 = [1,2], nums2 = [3,4]
 * Output: 2.50000
 * Explanation: merged array = [1,2,3,4] and median is (2 + 3) / 2 = 2.5.
 */

// Soluution : 

import java.util.*;

class Solution {
    public static double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int num1 = nums1.length, num2 = nums2.length;
        int[] ans = new int[num1 + num2];
        int i=0, j=0, k=0;

        // Insert elements in new array, until one or both of them reaches to last index.

        while(i<num1 && j<num2) {
            if(nums1[i] <= nums2[j])
                ans[k++] = nums1[i++];
            else
                ans[k++] = nums2[j++];
        } 

        // If nums2 reaches to last index.

        while(i < num1) {
            ans[k++] = nums1[i++];
        }

        // If num1 reaches to last index.

        while(j < num2) {
            ans[k++] = nums2[j++];
        }
        
        // Find out the median of sorted array by clarifying even/odd length.

        if((num1 + num2)%2 == 0)
            return (double)(ans[ans.length/2-1] + ans[ans.length/2])/2; 
        
        return ans[ans.length/2];
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n1 = sc.nextInt();
        int[] nums1 = new int[n1];

        for(int i=0; i<n1; i++)
            nums1[i] = sc.nextInt();
        
        int n2 = sc.nextInt();
        int[] nums2 = new int[n2];

        for(int i=0; i <n2; i++)
            nums2[i] = new int[n2];

        System.out.println( findMedianSortedArrays(nums1, nums2) );

        sc.close();
    }
}

// Time complexity : O(m + n)
// Space complexity : O(1)

File: LeetCode/Median of Two Sorted Arrays/Median of Two Sorted Arrays.cpp
#include <iostream>
#include <vector>
  
using namespace std;
// https://leetcode.com/problems/median-of-two-sorted-arrays/

// Method to find median of two sorted arrays(vectors) 
double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
    int m=nums1.size();
    int n=nums2.size();
    vector<float> merged;
    int i=0;
    int j=0;
   
    // Add elements in merged vector until one of the two vector get completed
    while(i<m && j<n){
        if(nums1[i]>nums2[j]){
            merged.push_back(nums2[j]) ;
            j++;
        }else{
            merged.push_back(nums1[i]);
            i++;
        }
    }
    // add remaining elements of nums2 if nums1 fully added
    if(j<n){
        while(j<n){
            merged.push_back(nums2[j]);
            j++;
        }
    }
    // add remaining elements of nums1 if nums2 fully added
        if(i<m){
        while(i<m){
            merged.push_back(nums1[i]);
            i++;
        }
    }
    
    
    if((m+n)%2!=0){
        return merged[(m+n-1)/2];
    }else{
        return  (merged[(m+n)/2]+merged[((m+n)/2)-1])/2;
    }
}



int main()
{
    vector<int> nums1 = {1, 3};
    vector<int> nums2 = {2};
    cout << findMedianSortedArrays(nums1, nums2);
}